Add discovery and cluster formation topic (#11416)
opensearch-project/documentation-website: _tuning-your-cluster/discovery-cluster-formation/settings.md

---
layout: default
title: Discovery and cluster formation settings
parent: Discovery and cluster formation
nav_order: 100
---

# Discovery and cluster formation settings

This page provides a comprehensive reference for all settings that control discovery and cluster formation behavior in OpenSearch. These settings determine how nodes find each other, elect a cluster manager, and maintain cluster coordination.

## Core discovery settings

The following settings control the fundamental discovery process:

> - `discovery.seed_hosts` (Static, list): Provides a list of addresses for cluster-manager-eligible nodes in the cluster. This setting is essential for nodes to find each other during cluster formation. Each address can be specified as `host:port` or just `host`. The `host` can be a hostname (resolved via DNS - if multiple IPs are resolved, OpenSearch attempts to connect to all), IPv4 address, or IPv6 address (must be enclosed in square brackets). If no port is specified, OpenSearch determines the port by checking `transport.profiles.default.port`, then `transport.port`. If neither is configured, the default port `9300` is used. Default is `["127.0.0.1", "[::1]"]`.

Reviewer: ```suggestion
- `discovery.seed_hosts` (Static, list): Provides a list of addresses for cluster-manager-eligible nodes in the cluster. This setting is essential for nodes to be able to find each other during cluster formation. Each address can be specified as `host:port` or just `host`. The `host` can be a hostname (resolved through DNS---if multiple IPs are resolved, OpenSearch attempts to connect to all), IPv4 address, or IPv6 address (must be enclosed in square brackets). If no port is specified, OpenSearch determines the port by checking `transport.profiles.default.port`, then `transport.port`. If neither is configured, the default port `9300` is used. Default is `["127.0.0.1", "[::1]"]`.
```

> - `discovery.seed_providers` (Static, list): Specifies which seed hosts providers to use for obtaining seed node addresses during discovery. Available providers are `settings` (uses addresses from the `discovery.seed_hosts` setting) and `file` (reads addresses from the `unicast_hosts.txt` file). You can specify multiple providers to combine discovery methods. Default is `["settings"]`.

Reviewer: ```suggestion
- `discovery.seed_providers` (Static, list): Specifies which seed host providers to use for obtaining seed node addresses during discovery. Available providers are `settings` (uses addresses from the `discovery.seed_hosts` setting) and `file` (reads addresses from the `unicast_hosts.txt` file). You can specify multiple providers to combine discovery methods. Default is `["settings"]`.
```

> - `discovery.type` (Static, string): Specifies whether OpenSearch should form a multiple-node cluster or operate as a single node. When set to `single-node`, OpenSearch forms a single-node cluster and suppresses certain timeouts. This setting is useful for development and testing environments. Valid values are `multi-node` (default) and `single-node`.

Reviewer: ```suggestion
- `discovery.type` (Static, string): Specifies whether OpenSearch should form a multi-node cluster or operate as a single node. When set to `single-node`, OpenSearch forms a single-node cluster and suppresses certain timeouts. This setting is useful for development and testing environments. Valid values are `multi-node` (default) and `single-node`.
```

- `cluster.initial_cluster_manager_nodes` (Static, list): Sets the initial cluster-manager-eligible nodes for bootstrapping a brand-new cluster. This setting is required when bootstrapping a cluster for the first time and should contain the node names (as defined by `node.name`) of the initial cluster-manager-eligible nodes. This list should be empty for nodes joining an existing cluster. Default is `[]` (empty).

> ## Discovery process settings

Reviewer 2: We can go with existing heading of `Expert Level Settings` as we don't expect process and dns settings to be tuned by user.

- [Leader checks](#leader-checks) (sent by non-cluster-manager nodes to the cluster manager)

OpenSearch allows occasional check failures and uses the following guidelines for taking action:

- Transient issues (single check failures) are ignored; multiple consecutive failures are required for action.
- Network disconnects trigger immediate response.
- All timeouts and retry counts are configurable.

### Follower checks 

The elected cluster manager periodically checks each node in the cluster:

1. Sends periodic health check requests to all nodes.
2. Waits for responses within the configured timeout.
3. Tracks consecutive check failures for each node.
4. Removes nodes that fail consecutive checks (based on retry count).

If the cluster manager detects a node has disconnected (network-level disconnect), it bypasses the timeout and retry settings and immediately attempts to remove the node from the cluster.

### Leader checks

Each non-cluster-manager node periodically checks the health of the elected cluster manager:

1. Send periodic health check requests to the cluster manager.
2. Wait for responses within the configured timeout.

> 3. Track consecutive check failures for the cluster manager.

Reviewer: ```suggestion
3. Tracks consecutive check failures for the cluster manager.
```

> 4. Start new cluster manager election if consecutive checks fail.

Reviewer: ```suggestion
4. Starts a new cluster manager election if consecutive checks fail.
```

> If a node detects the cluster manager has disconnected, it bypasses timeout and retry settings and immediately restarts its discovery phase to find or elect a new cluster manager.

Reviewer: ```suggestion
If a node detects that the cluster manager has disconnected, it bypasses timeout and retry settings and immediately restarts its discovery phase to find or elect a new cluster manager.
```

## Cluster state publishing settings

The following settings control how cluster state updates are distributed:

- `cluster.publish.timeout` (Static, time unit): Sets how long the cluster manager waits for cluster state updates to be published to all nodes before timing out. This setting is ignored when `discovery.type` is set to `single-node`. Default is `30s`.

- `cluster.publish.info_timeout` (Static, time unit): Sets how long the cluster manager waits before logging a message about slow-responding nodes during cluster state publishing. Default is `10s`.

- `cluster.follower_lag.timeout` (Static, time unit): Sets how long the cluster manager waits for acknowledgments of cluster state updates from lagging nodes. Nodes that don't respond within this time are considered failed and removed from the cluster. Default is `90s`.

## Cluster coordination settings

The following settings control cluster joining and coordination:

- `cluster.join.timeout` (Static, time unit): Sets how long a node waits after sending a join request before considering it failed and retrying. This setting is ignored when `discovery.type` is set to `single-node`. Default is `60s`.

> - `cluster.no_cluster_manager_block` (Dynamic, string): Specifies which operations are rejected when there is no active cluster manager. Valid values are `all` (all operations including read/write and cluster state APIs are rejected) and `write` (only write operations are rejected; read operations succeed based on the last known cluster state but may return stale data). This setting doesn't affect node-based APIs (cluster stats, node info, node stats). For full cluster functionality, an active cluster manager is required. Default is `write`.

Reviewer: ```suggestion
- `cluster.no_cluster_manager_block` (Dynamic, string): Specifies which operations are rejected when there is no active cluster manager. Valid values are `all` (all operations including read/write and cluster state APIs are rejected) and `write` (only write operations are rejected; read operations succeed based on the last known cluster state but may return stale data). This setting doesn't affect node-based APIs (Cluster Stats, Node Info, Node Stats). For full cluster functionality, an active cluster manager is required. Default is `write`.
```

  - cluster-manager-1
  - cluster-manager-2
  - cluster-manager-3

# Voting configuration
cluster.auto_shrink_voting_configuration: true
cluster.max_voting_config_exclusions: 10
```
{% include copy.html %}

### Development single-node setup

```yaml
# Single-node development cluster
cluster.name: dev-cluster
discovery.type: single-node

# Optional: Disable cluster formation timeouts for faster startup
cluster.publish.timeout: 5s
cluster.join.timeout: 10s
```
{% include copy.html %}

### High-availability production cluster

```yaml
# Production cluster with dedicated cluster manager nodes
cluster.name: production-cluster

# Discovery configuration
discovery.seed_hosts:
  - cluster-manager-1.example.com:9300
  - cluster-manager-2.example.com:9300
  - cluster-manager-3.example.com:9300

# Bootstrap configuration (only for initial setup)
cluster.initial_cluster_manager_nodes:
  - cluster-manager-1
  - cluster-manager-2
  - cluster-manager-3

# Production-optimized settings
cluster.auto_shrink_voting_configuration: true
cluster.max_voting_config_exclusions: 3
cluster.publish.timeout: 60s
cluster.join.timeout: 120s
```
{% include copy.html %}

## Related documentation

> - [Node discovery and seed hosts]({{site.url}}{{site.baseurl}}/tuning-your-cluster/discovery-cluster-formation/discovery/): Learn about discovery mechanisms and seed host providers

Reviewer: ```suggestion
- [Node discovery and seed hosts]({{site.url}}{{site.baseurl}}/tuning-your-cluster/discovery-cluster-formation/discovery/): Provides information about discovery mechanisms and seed host providers
```
